core/src/com/unciv/logic/BarbarianManager.kt

```diff
@@ -0,0 +1,269 @@
+        return toReturn
+    }
+
+    fun update() {
+        if (countdown <= 0) {
+            // Try spawning a unit
+            if (spawnBarbarian()) {
+                // Successful
+                spawnedUnits++
+                resetCountdown()
+            }
+        } else
+            countdown--
+    }
+
+    fun wasAttacked() {
+        countdown /= 2
+    }
+
+    /** Attempts to spawn a Barbarian from this encampment. Returns true if a unit was spawned. */
+    private fun spawnBarbarian(): Boolean {
+        val barbarianCiv = gameInfo.getBarbarianCivilization()
+        val tile = gameInfo.tileMap[position]
+
+        // Empty camp - spawn a defender
+        if (tile.militaryUnit == null) {
+            val spawnedUnit = tile.tileMap.placeUnitNearTile(tile.position, chooseBarbarianUnit(naval = false), barbarianCiv)
+            return (spawnedUnit != null)
+        }
+
+        // Don't spawn wandering barbs too early
+        if (gameInfo.turns < 10)
+            return false
+
+        // Too many barbarians around already?
+        if (tile.getTilesInDistance(4).count { it.militaryUnit?.civInfo == barbarianCiv } > 2)
+            return false
+
+        val canSpawnBoats = gameInfo.turns > 30
+        val validTiles = tile.neighbors.toList().filterNot {
+            it.isImpassible()
+                    || it.isCityCenter()
+                    || it.getFirstUnit() != null
+                    || (it.isWater && !canSpawnBoats)
+                    || (it.hasUnique("Fresh water") && it.isWater) // No Lakes
+        }
+        if (validTiles.isEmpty()) return false
+        val spawnTile = validTiles.random()
+
+        val spawnedUnit = gameInfo.tileMap.placeUnitNearTile(spawnTile.position, chooseBarbarianUnit(spawnTile.isWater), barbarianCiv)
+        return (spawnedUnit != null)
+    }
+
+    private fun chooseBarbarianUnit(naval: Boolean): String {
+        // if we don't make this into a separate list then the retain() will happen on the Tech keys,
+        // which effectively removes those techs from the game and causes all sorts of problems
+        val allResearchedTechs = gameInfo.ruleSet.technologies.keys.toMutableList()
+        for (civ in gameInfo.civilizations.filter { !it.isBarbarian() && !it.isDefeated() }) {
+            allResearchedTechs.retainAll(civ.tech.techsResearched)
+        }
+        val barbarianCiv = gameInfo.getBarbarianCivilization()
+        barbarianCiv.tech.techsResearched = allResearchedTechs.toHashSet()
+        val unitList = gameInfo.ruleSet.units.values
+            .filter { it.isMilitary() }
+            .filter { it.isBuildable(barbarianCiv) }
+
+        var unit = if (naval)
+            unitList.filter { it.isWaterUnit() }.randomOrNull()
+        else
+            unitList.filter { it.isLandUnit() }.randomOrNull()
+
+        if (unit == null) // Didn't find a unit for preferred domain
+            unit = unitList.random() // Pick another, or crash the game vOv
```

Comment:
How about adding if unitlist isempty return on line 237 to defend against mad modders

Reply:
randomOrNull should exist as well